rottingOranges.js

```diff
@@ -0,0 +1,150 @@
+/*
+We're going to approach this problem with a multi-source BFS where can gather the location of a
+rotten orange, set it with a starting time of 0, and add it to the queue. We can also gather the total
+number of fresh oranges in the grid and keep track of them as we traverse/BFS-ing.
+
+Once in the queue, we can traverse the grid, perform a BFS until we find a fresh orange
+and update the current time so far but only if it's the first time visiting that fresh orange. We will
+use a visited set to keep track of all the fresh oranges visited.
+
+As we're BFS-ing the grid, we can keep track of the number of fresh oranges and decrese it by 1 if it's
+our first time visiting that fresh orange. Once we reach 0 fresh oranges, we can return the current time.
+*/
+
```

Comment:
Since we're trying to find the minimum number of minutes that pass for all fresh oranges to become rotten, using a multi-sourced BFS would be ideal because starting from the source (in this case the rotten oranges from multiple locations), we can set our starting time to be 0.

And by using a visited set, we can keep track of all the locations we've already visited and only update the time if we've visiting a fresh orange for he first time. When visiting a fresh orange for the first time, it will be shortest path thus giving us the minimum time.

Reply:
You didn't really tell me anything unique about this problem that requires a multi-sourced BFS. From the description you gave, it sounds like a normal BFS would be enough. Give this another shot.

Reply:
Let me know what you think:

It's because we're trying to find the minimum number of minutes it takes for all fresh oranges to become rotten. I believe the keyword here is minimum. And since there are multi-rotten oranges scattered around the gird, along with fresh oranges, using a multi-source BFS would be a good choice because some rotten oranges will be closer to fresh ones at the start.

Reply:
Not exactly, the word minimum tells us that we need to use a BFS. But the important question is why are we using a multi-sourced BFS. The answer is because we have things happening in parallel (i.e. at the same time). Suppose we have two rotten oranges on opposite corners. Then, at the exact same time, they are BOTH rotting their neighbors. A normal BFS wouldn't allow us to update all the relevant squares at the same time. We'd have to first do all the rotting for orange #1, and then do it for orange #2. But, because we have this rotting happening in parallel, we must use a multi-sourced BFS.

Now that you've seen my answer, I suggest that you read it carefully and jot down some notes. Then, try to write up an answer without re-reading my answer. Feel free to use your notes.

Reply:
Ah, that makes sense. Let me know what you think of explanation:

It's because the grid contains various locations of rotten oranges in contact with fresh ones and the goal is to find the minimum number of minutes for all fresh ones to become rotten. Since all fresh oranges are becoming rotten simultaneously, using a multi-source BFS would be ideal because we're able to gather all the locations of the rotten oranges at the start and only update the time when we visit a fresh orange for the first time.

Updating the time will take account for other all other fresh oranges as well and we'll keep track of all the visited locations with a visited set.

Reply:
Much better!